### PR TITLE
GH-2478: Made the Java LS download URL more flexible.

### LINKED ISSUE #2478: [java] Cannot customize the download link of the Java LS
Although one can specify the `ls.download.path` and the `ls.download.base` in the `package.json` but there is a hard-coded path (`&r=1`) part of the URL.

https://github.com/theia-ide/theia/blob/c5554c8a7bfe909ca04c7bde6c95070decaf3adf/packages/java/scripts/download-jdt-ls.js#L31

Also, later we make additional assumptions from the query part of the URL.

https://github.com/theia-ide/theia/blob/c5554c8a7bfe909ca04c7bde6c95070decaf3adf/packages/java/scripts/download-jdt-ls.js#L32

Acceptance criteria:
There should be a way to download the Java LS from any other URLs. For instance:
`https://github.com/user-name/repositor-name/raw/branch-name/jdt.ls.tar.gz`.

Otherwise, the declaration of the LS URL in the `package.json` works for the happy path but not for others.

### PR DESCRIPTION
**Pull request by kittaakos**
Introduced a download history JSON file to track the
original file source (download URL) with the SHA1 hash
of the file name. So there is no need to hard-code
or assume anything about the file names.

Fix the broken property access of the download URL.

Closes #2478.

Signed-off-by: Akos Kitta <kittaakos@typefox.io>

### CODE REVIEW

**Comment by kittaakos** [comment]
@AlexTugarev, please ping me if you do not have the time for this review. Thanks!

File: packages/java/scripts/download-jdt-ls.js [hunk 1]
```diff
@@ -47,32 +51,42 @@ function downloadJavaServer() {
...
                 }
             })
 
         };
-        downloadWithRedirect(archiveUri);
+        downloadWithRedirect(downloadUrl);
     });
 }
 
+// Just to make sure we can reverse-engineer the download URL from the hash of the file name.
```
Comment (AlexTugarev) [inline packages/java/scripts/download-jdt-ls.js]: When do we need this info? And would a simple log file also work to just understand the download history?

File: packages/java/package.json [hunk 2]
```diff
@@ -57,7 +58,6 @@
     "extends": "../../configs/nyc.json"
   },
   "ls": {
-    "download.base": "https://www.eclipse.org/downloads/download.php?file=",
-    "download.path": "/che/che-ls-jdt/snapshots/che-jdt-language-server-latest.tar.gz"
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/java/package.json]: > case for separate properties at all

It was bogus anyway and always resolved to `undefine` so the fallback was used.

File: packages/java/scripts/download-jdt-ls.js [hunk 3]
```diff
@@ -47,32 +51,42 @@ function downloadJavaServer() {
         const file = fs.createWriteStream(archivePath);
         const downloadWithRedirect = url => {
             /** @type { any } */
-            const h = url.toString().startsWith("https") ? https : http;
+            const h = url.toString().startsWith('https') ? https : http;
             h.get(url, response => {
                 const statusCode = response.statusCode;
                 const redirectLocation = response.headers.location;
                 if (statusCode >= 300 && statusCode < 400 && redirectLocation) {
-                    console.log("redirect location: " + redirectLocation)
```
[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/java/scripts/download-jdt-ls.js]: Why do you consider it informative when it does not show up in the logs after calling `yarn`? Am I missing something?

File: packages/java/scripts/download-jdt-ls.js (hunk 1 above)
Comment (kittaakos) [reply packages/java/scripts/download-jdt-ls.js]: I do not have any preferences on that. Eventually, we can get rid of this download history at all.

File: packages/java/scripts/download-jdt-ls.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/java/scripts/download-jdt-ls.js]: Anyway, I'll restore it.

[the remark this answers is not in the document]
Comment (AlexTugarev) [reply packages/java/scripts/download-jdt-ls.js]: No it doesn't show up in the logs, but it's printed to stdout same way as it's done for build of the filesystem watcher lib. Once you face a slow eclipse mirror site, you'll notice.

File: packages/java/package.json (hunk 2 above)
[the remark this answers is not in the document]
Comment (AlexTugarev) [reply packages/java/package.json]: Well in that case 👍

File: packages/java/scripts/download-jdt-ls.js (hunk 1 above)
Comment (AlexTugarev) [reply packages/java/scripts/download-jdt-ls.js]: > I do not have any preferences on that.

Me neither. Just asking for a reason. But if that can help debugging...

File: packages/java/scripts/download-jdt-ls.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (kittaakos) [reply packages/java/scripts/download-jdt-ls.js]: > Anyway, I'll restore it.

**Comment by kittaakos** [comment]
The Git watcher test failure (on Windows) is unrelated to my Java download link changes.

**Pull request merged**